fbgemm_gpu/src/dram_kv_embedding_cache/feature_evict.h

```diff
@@ -0,0 +1,199 @@
+//
+// Created by root on 25-5-26.
+//
+#pragma once
+
+#include <atomic>
+#include <chrono>
+#include <cstddef>
+#include <memory_resource>
+#include <stdexcept>
+#include <vector>
+
+#include <cassert>
+#include <folly/executors/CPUThreadPoolExecutor.h>
+#include <folly/futures/Future.h>
+
+#include "SynchronizedShardedMap.h"
+
+namespace kv_mem {
+
+class FeatureEvictBase {
+ public:
+  FeatureEvictBase(folly::CPUThreadPoolExecutor* executor,
+                   SynchronizedShardedMap<int64_t, float*>& kv_store)
+      : executor_(executor),
+        kv_store_(kv_store),
+        evict_flag_(false),
+        evict_interrupt_(false),
+        num_shards_(kv_store.getNumShards()) {
+    init_shard_status();
+    // evict_flag_ 表示是否有任务在进行
+    // evict_interrupt_ 表示是否有任务被中断
+  }
+
+  virtual ~FeatureEvictBase() {
+    // 析构时，需要等待任务执行完成
+    wait_completion();  // 等待所有异步任务完成
+  };
+
+  // 触发异步淘汰
+  // 如果有执行中的任务，直接返回, 防止多次触发
+  // 如果没有执行中的任务，初始化任务状态
+  void trigger_evict() {
+    std::lock_guard<std::mutex> lock(mutex_);
+    if (evict_flag_.exchange(true)) return;
+    prepare_evict();
+  }
+
+  // 恢复任务执行，如果有进行中的任务返回true, 没有返回false
+  bool resume() {
+    std::lock_guard<std::mutex> lock(mutex_);
+    if (!evict_flag_.load()) return false;
+    evict_interrupt_.store(false);
+    for (int shard_id = 0; shard_id < num_shards_; ++shard_id) {
+      submit_shard_task(shard_id);
+    }
+    return true;
+  };
+
+  // 暂停淘汰过程，如果有进行中的任务返回true, 没有返回false
+  // 在暂停阶段，判断淘汰是否完成
+  bool pause() {
+    std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
is this lock required?

Reply:
We hope that feature elimination is thread-safe and not affected by the calling method. At present, I understand that all interfaces may not be guaranteed to be safe. For example, trigger may be called by one thread, and pause is called by another thread, so locking is required.